File: apps/androidtv/core/corekit/src/main/kotlin/dev/spidola/tv/core/corekit/ActionableError.kt
// SPDX-FileCopyrightText: 2026 Spidola contributors
// SPDX-License-Identifier: AGPL-3.0-or-later

package dev.spidola.tv.core.corekit

import android.content.Context
import uniffi.core_api.ApiException
import uniffi.core_api.InputField
import uniffi.core_api.InputIssue

/**
 * A prescribed user action for an error (PRD §6.3). The set is deliberately small; the shell
 * renders each as a focusable button.
 */
enum class ErrorAction(
    val label: Int,
) {
    RETRY(R.string.core_action_retry),
    GO_BACK(R.string.core_action_go_back),
    FIX_INPUT(R.string.core_action_edit),
}

/** A shell-owned string resource with an optional shell-owned string argument. */
data class LocalizedText(
    val resource: Int,
    val argument: Int? = null,
) {
    /** Resolves the resource only at the Android presentation edge. */
    fun resolve(context: Context): String =
        argument?.let { context.getString(resource, context.getString(it)) } ?: context.getString(resource)
}

/**
 * The plain-language presentation of an [ApiException]: a short failure class, a one-sentence
 * message, and a **non-empty** set of prescribed actions (PRD §6.3, mirroring the core's
 * `ApiError::ux` table in `crates/core-api/src/error.rs`). Diagnostic detail stays in the log
 * stream, never here (PRD §8.6).
 *
 * "No action available" is unrepresentable: [primaryAction] is a single required value, so every
 * `ActionableError` carries at least one action by construction — a UI that renders one can never
 * be handed an actionless error.
 */
data class ActionableError(
    val failureClass: LocalizedText,
    val message: LocalizedText,
    val primaryAction: ErrorAction,
    val otherActions: List<ErrorAction>,
) {
    /** Every offered action, primary first — always non-empty. */
    val actions: List<ErrorAction> get() = listOf(primaryAction) + otherActions

    companion object {
        /**
         * Maps a boundary [ApiException] onto its presentation. The `when` is exhaustive over the
         * sealed error hierarchy; a variant added to the core forces a decision here at compile
         * time (TECH_SPEC §5).
         */
        fun from(error: ApiException): ActionableError =
            when (error) {
                is ApiException.NetworkUnreachable ->
                    ActionableError(
                        LocalizedText(R.string.core_error_network_title),
                        LocalizedText(R.string.core_error_network_message),
                        ErrorAction.RETRY,
                        listOf(ErrorAction.GO_BACK),
                    )
                is ApiException.Timeout ->
                    ActionableError(
                        LocalizedText(R.string.core_error_timeout_title),
                        LocalizedText(R.string.core_error_timeout_message),
                        ErrorAction.RETRY,
                        listOf(ErrorAction.GO_BACK),
                    )
                is ApiException.Unauthorized ->
                    ActionableError(
                        LocalizedText(R.string.core_error_unauthorized_title),
                        LocalizedText(R.string.core_error_unauthorized_message),
                        ErrorAction.FIX_INPUT,
                        listOf(ErrorAction.GO_BACK),
                    )
                is ApiException.NotFound ->
                    ActionableError(
                        LocalizedText(R.string.core_error_not_found_title),
                        LocalizedText(R.string.core_error_not_found_message),
                        ErrorAction.GO_BACK,
                        emptyList(),
                    )
                is ApiException.InvalidInput ->
                    ActionableError(
                        LocalizedText(R.string.core_error_invalid_title),
                        inputMessage(error.field, error.issue),
                        ErrorAction.FIX_INPUT,
                        listOf(ErrorAction.GO_BACK),
                    )
                is ApiException.ParseFailed ->
                    ActionableError(
                        LocalizedText(R.string.core_error_parse_title),
                        LocalizedText(R.string.core_error_parse_message),
                        ErrorAction.RETRY,
                        listOf(ErrorAction.GO_BACK),
                    )
                is ApiException.StorageCorrupt ->
                    ActionableError(
                        LocalizedText(R.string.core_error_storage_title),
                        LocalizedText(R.string.core_error_storage_message),
                        ErrorAction.RETRY,
                        listOf(ErrorAction.GO_BACK),
                    )
                is ApiException.Cancelled ->
                    ActionableError(
                        LocalizedText(R.string.core_error_cancelled_title),
                        LocalizedText(R.string.core_error_cancelled_message),
                        ErrorAction.GO_BACK,
                        emptyList(),
                    )
                is ApiException.Internal ->
                    ActionableError(
                        LocalizedText(R.string.core_error_internal_title),
                        LocalizedText(R.string.core_error_internal_message),
                        ErrorAction.RETRY,
                        listOf(ErrorAction.GO_BACK),
                    )
            }

        /** Turns stable boundary codes into shell-owned copy. The generated exception's message is
         * intentionally ignored because it is diagnostic data, not localized UI prose. */
        private fun inputMessage(
            field: InputField,
            issue: InputIssue,
        ): LocalizedText {
            val fieldResource =
                when (field) {
                    InputField.ADDRESS -> R.string.core_error_field_address
                    InputField.SERVER -> R.string.core_error_field_server
                    InputField.NAME -> R.string.core_error_field_name
                    InputField.HEADER -> R.string.core_error_field_header
                    InputField.LOG_LEVEL -> R.string.core_error_field_log_level
                    InputField.FILE -> R.string.core_error_field_file
                    InputField.SOURCE -> R.string.core_error_field_source
                }
            val messageResource =
                when (issue) {
                    InputIssue.EMPTY -> R.string.core_error_issue_empty
                    InputIssue.INVALID -> R.string.core_error_issue_invalid
                    InputIssue.UNSUPPORTED -> R.string.core_error_issue_unsupported
                    InputIssue.UNAVAILABLE -> R.string.core_error_issue_unavailable
                }
            return LocalizedText(messageResource, fieldResource)
        }
    }
}

File: apps/tvos/Packages/CoreKit/Sources/CoreKit/ActionableError.swift
// SPDX-FileCopyrightText: 2026 Spidola contributors
// SPDX-License-Identifier: AGPL-3.0-or-later

import Foundation
import core_api

/// A prescribed user action for an error (PRD §6.3). The set is deliberately small; the shell
/// renders each as a focusable button.
public enum ErrorAction: Sendable, Hashable {
  /// Try the same operation again.
  case retry
  /// Return to the previous screen.
  case goBack
  /// Correct the input that caused the failure.
  case fixInput

  /// The couch-legible button label (PRD §8.6 voice).
  public var label: String {
    switch self {
    case .retry: String(localized: "Try again", bundle: .module)
    case .goBack: String(localized: "Go back", bundle: .module)
    case .fixInput: String(localized: "Edit", bundle: .module)
    }
  }
}

/// The plain-language presentation of an `ApiError`: a short failure class, a one-sentence
/// message, and a **non-empty** set of prescribed actions (PRD §6.3, mirroring the core's
/// `ApiError::ux` table in `crates/core-api/src/error.rs`). Diagnostic detail stays in the log
/// stream, never here (PRD §8.6).
///
/// "No action available" is unrepresentable: `primaryAction` is a single required value, so every
/// `ActionableError` carries at least one action by construction — a UI that renders one can never
/// be handed an actionless error.
public struct ActionableError: Sendable, Equatable {
  /// A short, couch-legible failure class.
  public let failureClass: String
  /// A one-sentence, jargon-free explanation of what happened.
  public let message: String
  /// The recommended action, always present.
  public let primaryAction: ErrorAction
  /// Further actions offered after the primary one.
  public let otherActions: [ErrorAction]

  /// Every offered action, primary first — always non-empty.
  public var actions: [ErrorAction] { [primaryAction] + otherActions }

  private init(
    _ failureClass: String,
    _ message: String,
    primary: ErrorAction,
    other: [ErrorAction]
  ) {
    self.failureClass = failureClass
    self.message = message
    self.primaryAction = primary
    self.otherActions = other
  }

  /// Maps a boundary `ApiError` onto its presentation. The `@unknown default` reserves the
  /// "unknown future variant" arm the FFI boundary rules require (TECH_SPEC §5), so an older shell
  /// still presents a newer core's error with an action rather than crashing.
  public init(_ error: ApiError) {
    switch error {
    case .NetworkUnreachable:
      self.init(
        String(localized: "Can't reach the source", bundle: .module),
        String(
          localized:
            "Spidola couldn't connect. Check the address and your network, then try again.",
          bundle: .module),
        primary: .retry, other: [.goBack])
    case .Timeout:
      self.init(
        String(localized: "The source is slow to respond", bundle: .module),
        String(
          localized: "The source didn't answer in time. It may be busy — try again in a moment.",
          bundle: .module),
        primary: .retry, other: [.goBack])
    case .Unauthorized:
      self.init(
        String(localized: "Login was rejected", bundle: .module),
        String(
          localized: "The source didn't accept these sign-in details. Edit them and try again.",
          bundle: .module),
        primary: .fixInput, other: [.goBack])
    case .NotFound:
      self.init(
        String(localized: "Not available anymore", bundle: .module),
        String(localized: "This isn't at the source any longer.", bundle: .module),
        primary: .goBack, other: [])
    case .InvalidInput(let field, let issue):
      self.init(
        String(localized: "That entry isn't valid", bundle: .module),
        Self.invalidInputMessage(field: field, issue: issue),
        primary: .fixInput, other: [.goBack])
    case .ParseFailed:
      self.init(
        String(localized: "No channels found", bundle: .module),
        String(
          localized:
            "Spidola reached the source but found no channels to add. Check the playlist and try again.",
          bundle: .module),
        primary: .retry, other: [.goBack])
    case .StorageCorrupt:
      self.init(
        String(localized: "Local storage problem", bundle: .module),
        String(
          localized: "Something went wrong saving to this device. Try again.", bundle: .module),
        primary: .retry, other: [.goBack])
    case .Cancelled:
      self.init(
        String(localized: "Cancelled", bundle: .module),
        String(localized: "That was cancelled.", bundle: .module),
        primary: .goBack, other: [])
    case .Internal:
      self.init(
        String(localized: "Something went wrong", bundle: .module),
        String(localized: "An unexpected problem came up. Try again.", bundle: .module),
        primary: .retry, other: [.goBack])
    @unknown default:
      self.init(
        String(localized: "Something went wrong", bundle: .module),
        String(localized: "An unexpected problem came up. Try again.", bundle: .module),
        primary: .retry, other: [.goBack])
    }
  }

  /// Platform copy for the structured validation boundary. No core-authored prose crosses into UI.
  private static func invalidInputMessage(field: InputField, issue: InputIssue) -> String {
    let fieldName: String =
      switch field {
      case .address: String(localized: "address", bundle: .module)
      case .server: String(localized: "server address", bundle: .module)
      case .name: String(localized: "name", bundle: .module)
      case .header: String(localized: "request detail", bundle: .module)
      case .logLevel: String(localized: "log level", bundle: .module)
      case .file: String(localized: "file", bundle: .module)
      case .source: String(localized: "source", bundle: .module)
      @unknown default: String(localized: "entry", bundle: .module)
      }

    let format: String =
      switch issue {
      case .empty: String(localized: "Enter a value for the %@.", bundle: .module)
      case .invalid: String(localized: "Check the %@ and try again.", bundle: .module)
      case .unsupported: String(localized: "That %@ isn't supported.", bundle: .module)
      case .unavailable: String(localized: "That %@ isn't available right now.", bundle: .module)
      @unknown default: String(localized: "Check the %@ and try again.", bundle: .module)
      }
    return String(format: format, locale: .current, fieldName)
  }
}
